### PR TITLE
Full user profile access fix

### PR DESCRIPTION
**Pull request by hryhoriivoloshyn**
dev
## JIRA

* [Main JIRA ticket](https://jira.softserve.academy/secure/RapidBoard.jspa?rapidView=id)

## Code reviewers

- [ ] @github_username

### Second Level Review

- [ ] @github_username

## Summary of issue

ToDo

## Summary of change

ToDo

## Testing approach

ToDo

## CHECK LIST
- [ ]  СI passed
- [ ]  Сode coverage >=95%
- [ ]  PR is reviewed manually again (to make sure you have 100% ready code)
- [ ]  All reviewers agreed to merge the PR
- [ ]  I've checked new feature as logged in and logged out user if needed
- [ ]  PR meets all conventions

### CODE REVIEW

**Review by vsynieok — APPROVED** [review]

File: EPlast/EPlast.WebApi/Controllers/UserController.cs [hunk 1]
```diff
@@ -138,51 +141,45 @@ public async Task<IActionResult> GetUserProfile(string currentUserId, string foc
...
-            if (focusUser == null)
+            var currentUser = await _userManager.FindByIdAsync(currentUserId);
+
+            var currentUserAccess = await 
+                _userAccessService.GetUserProfileAccessAsync(currentUserId, focusUserId, currentUser);
+            if (currentUserAccess == null)
             {
-                _loggerService.LogError($"User not found. UserId:{focusUserId}");
-                return NotFound();
+                return StatusCode(StatusCodes.Status500InternalServerError);
```
Comment (tvardero) [inline EPlast/EPlast.WebApi/Controllers/UserController.cs:151]: Do not throw 500 without any comment 

Also, better use `throw new ....Exception()`, it's more descriptive

File: EPlast/EPlast.WebApi/Controllers/UserController.cs [hunk 2]
```diff
@@ -132,20 +132,20 @@ public async Task<IActionResult> PutComment(string userId, [FromBody] string com
...
+        public async Task<IActionResult> GetUserProfile(string focusUserId)
         {
             if (string.IsNullOrEmpty(focusUserId))
             {
                 _loggerService.LogError("User id is null");
                 return NotFound();
             }
 
-            var currentUser = await _userManager.FindByIdAsync(currentUserId);
+            var currentUser = await _userManager.GetUserAsync(HttpContext.User);
```
Comment (tvardero) [inline EPlast/EPlast.WebApi/Controllers/UserController.cs:145]: User is accessable without `HttpContext.`. Just `User`

File: EPlast/EPlast.WebApi/Controllers/UserController.cs (hunk 1 above)
Comment (hryhoriivoloshyn) [reply EPlast/EPlast.WebApi/Controllers/UserController.cs]: Ok, I decided to remove check for null altogether. Method GetUserProfileAccessAsync either will return filled dictionary or will throw an exception if json is not found, either way dictionary cannot be null.

File: EPlast/EPlast.WebApi/Controllers/UserController.cs [hunk 3]
```diff
@@ -139,23 +139,18 @@ public async Task<IActionResult> GetUserProfile(string focusUserId)
             if (string.IsNullOrEmpty(focusUserId))
```
Comment (tvardero) [inline EPlast/EPlast.WebApi/Controllers/UserController.cs:139]: `IsNullOrWhitespace` is better in most cases

File: EPlast/EPlast.WebApi/Controllers/UserController.cs [hunk 4]
```diff
@@ -139,23 +139,18 @@ public async Task<IActionResult> GetUserProfile(string focusUserId)
...
             {
                 _loggerService.LogError("User id is null");
+                return BadRequest("user id cannot be empty");
```
Comment (tvardero) [inline EPlast/EPlast.WebApi/Controllers/UserController.cs:142]: There are 2 common approaches:
1.
```c#
if(string.IsNullOrWhitespace(focusUserId)) ModelState.AddModelError(nameof(focusUserId), "Id can not be empty");
// ...
if (!ModelState.IsValid) return BadRequest(ModelState);
```

or
2. 
```c#
public async Task<IActionResult> GetUserProfile([Required] string focusUserId)
{ /* ... */ }
```

Comment (hryhoriivoloshyn) [reply EPlast/EPlast.WebApi/Controllers/UserController.cs]: Thanks, added [Required] attribute

File: EPlast/EPlast.WebApi/Controllers/UserController.cs (hunk 2 above)
Comment (hryhoriivoloshyn) [reply EPlast/EPlast.WebApi/Controllers/UserController.cs]: Removed this prefix

File: EPlast/EPlast.WebApi/Controllers/UserController.cs (hunk 3 above)
Comment (hryhoriivoloshyn) [reply EPlast/EPlast.WebApi/Controllers/UserController.cs]: Added Required attribute instead

**Review by tvardero — APPROVED** [review]

**Review by Serhii-Buniak — APPROVED** [review]

**Pull request merged**